File: bukkit/elysium-chat-bukkit/src/main/java/com/seventh_root/elysium/chat/bukkit/command/chatchannel/ChatChannelCreateCommand.java
package com.seventh_root.elysium.chat.bukkit.command.chatchannel;

import com.seventh_root.elysium.chat.bukkit.ElysiumChatBukkit;
import com.seventh_root.elysium.chat.bukkit.chatchannel.BukkitChatChannel;
import com.seventh_root.elysium.chat.bukkit.chatchannel.BukkitChatChannelProvider;
import org.bukkit.ChatColor;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.conversations.*;
import org.bukkit.entity.Player;

public class ChatChannelCreateCommand implements CommandExecutor {

    private final ElysiumChatBukkit plugin;
    private final ConversationFactory conversationFactory;

    public ChatChannelCreateCommand(ElysiumChatBukkit plugin) {
        this.plugin = plugin;
        conversationFactory = new ConversationFactory(plugin)
                .withModality(true)
                .withFirstPrompt(new ChatChannelNamePrompt())
                .withEscapeSequence("cancel")
                .thatExcludesNonPlayersWithMessage(ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.not-from-console")))
                .addConversationAbandonedListener(event -> {
                    if (!event.gracefulExit()) {
                        Conversable conversable = event.getContext().getForWhom();
                        if (conversable instanceof Player) {
                            Player player = (Player) conversable;
                            player.sendMessage(ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.operation-cancelled")));
                        }
                    }
                });
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (sender instanceof Conversable) {
            if (sender.hasPermission("elysium.chat.command.chatchannel.create")) {
                conversationFactory.buildConversation((Conversable) sender).begin();
            } else {
                sender.sendMessage(ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.no-permission-chatchannel-create")));
            }
        } else {
            sender.sendMessage(ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.not-from-console")));
        }
        return true;
    }

    private class ChatChannelNamePrompt extends ValidatingPrompt {

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-create-prompt"));
        }

        @Override
        protected boolean isInputValid(ConversationContext context, String input) {
            BukkitChatChannelProvider chatChannelProvider = plugin.getCore().getServiceManager().getServiceProvider(BukkitChatChannelProvider.class);
            return chatChannelProvider.getChatChannel(input) == null;
        }

        @Override
        protected String getFailedValidationText(ConversationContext context, String invalidInput) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-name-invalid-name"));
        }

        @Override
        protected Prompt acceptValidatedInput(ConversationContext context, String input) {
            context.setSessionData("name", input);
            return new ChatChannelNameSetPrompt();
        }

    }

    private class ChatChannelNameSetPrompt extends MessagePrompt {

        @Override
        protected Prompt getNextPrompt(ConversationContext context) {
            return new ChatChannelColorPrompt();
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-name-valid"));
        }

    }

    private class ChatChannelColorPrompt extends ValidatingPrompt {

        @Override
        protected boolean isInputValid(ConversationContext context, String input) {
            try {
                ChatColor.valueOf(input.toUpperCase().replace(' ', '_'));
                return true;
            } catch (IllegalArgumentException exception) {
                return false;
            }
        }

        @Override
        protected Prompt acceptValidatedInput(ConversationContext context, String input) {
            context.setSessionData("color", ChatColor.valueOf(input.toUpperCase().replace(' ', '_')));
            return new ChatChannelColorSetPrompt();
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-color-prompt"));
        }

        @Override
        protected String getFailedValidationText(ConversationContext context, String invalidInput) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-color-invalid-color"));
        }

    }

    private class ChatChannelColorSetPrompt extends MessagePrompt {

        @Override
        protected Prompt getNextPrompt(ConversationContext context) {
            return new ChatChannelFormatStringPrompt();
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-color-valid"));
        }

    }

    private class ChatChannelFormatStringPrompt extends StringPrompt {

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-formatstring-prompt"));
        }

        @Override
        public Prompt acceptInput(ConversationContext context, String input) {
            context.setSessionData("format_string", input);
            return new ChatChannelFormatStringSetPrompt();
        }

    }

    private class ChatChannelFormatStringSetPrompt extends MessagePrompt {

        @Override
        protected Prompt getNextPrompt(ConversationContext context) {
            return new ChatChannelRadiusPrompt();
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-formatstring-valid"));
        }

    }

    private class ChatChannelRadiusPrompt extends NumericPrompt {

        @Override
        protected Prompt acceptValidatedInput(ConversationContext context, Number input) {
            context.setSessionData("radius", input.intValue());
            return new ChatChannelRadiusSetPrompt();
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-radius-prompt"));
        }

    }

    private class ChatChannelRadiusSetPrompt extends MessagePrompt {

        @Override
        protected Prompt getNextPrompt(ConversationContext context) {
            if ((int) context.getSessionData("radius") > 0) {
                return new ChatChannelClearRadiusPrompt();
            } else {
                context.setSessionData("clear_radius", 0);
                return new ChatChannelMatchPatternPrompt();
            }
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-radius-valid"));
        }

    }

    private class ChatChannelClearRadiusPrompt extends NumericPrompt {

        @Override
        protected Prompt acceptValidatedInput(ConversationContext context, Number input) {
            context.setSessionData("clear_radius", input.intValue());
            return new ChatChannelClearRadiusSetPrompt();
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-clear-radius-prompt"));
        }

    }

    private class ChatChannelClearRadiusSetPrompt extends MessagePrompt {

        @Override
        protected Prompt getNextPrompt(ConversationContext context) {
            return new ChatChannelMatchPatternPrompt();
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-clear-radius-valid"));
        }

    }

    private class ChatChannelMatchPatternPrompt extends StringPrompt {

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-match-pattern-prompt"));
        }

        @Override
        public Prompt acceptInput(ConversationContext context, String input) {
            if (!input.equalsIgnoreCase("none"))
                context.setSessionData("match_pattern", input);
            else
                context.setSessionData("match_pattern", "");
            return new ChatChannelMatchPatternSetPrompt();
        }

    }

    private class ChatChannelMatchPatternSetPrompt extends MessagePrompt {

        @Override
        protected Prompt getNextPrompt(ConversationContext context) {
            if ((int) context.getSessionData("radius") <= 0) {
                return new ChatChannelIRCEnabledPrompt();
            } else {
                context.setSessionData("irc_enabled", false);
                context.setSessionData("irc_channel", "");
                context.setSessionData("irc_whitelist", false);
                return new ChatChannelJoinedByDefaultPrompt();
            }
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-match-pattern-valid"));
        }

    }

    private class ChatChannelIRCEnabledPrompt extends BooleanPrompt {

        @Override
        protected String getFailedValidationText(ConversationContext context, String invalidInput) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel.set-irc-enabled-invalid-boolean"));
        }

        @Override
        protected Prompt acceptValidatedInput(ConversationContext context, boolean input) {
            context.setSessionData("irc_enabled", input);
            return new ChatChannelIRCEnabledSetPrompt();
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-irc-enabled-prompt"));
        }

    }

    private class ChatChannelIRCEnabledSetPrompt extends MessagePrompt {

        @Override
        protected Prompt getNextPrompt(ConversationContext context) {
            if ((boolean) context.getSessionData("irc_enabled")) {
                return new ChatChannelIRCChannelPrompt();
            } else {
                context.setSessionData("irc_channel", "");
                context.setSessionData("irc_whitelist", false);
                return new ChatChannelJoinedByDefaultPrompt();
            }
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-irc-enabled-valid"));
        }

    }

    private class ChatChannelIRCChannelPrompt extends ValidatingPrompt {

        @Override
        protected boolean isInputValid(ConversationContext context, String input) {
            return input.matches("/([#&][^\\x07\\x2C\\s]{0,200})/");
        }

        @Override
        protected String getFailedValidationText(ConversationContext context, String invalidInput) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-irc-channel-invalid-irc-channel"));
        }

        @Override
        protected Prompt acceptValidatedInput(ConversationContext context, String input) {
            context.setSessionData("irc_channel", input);
            return new ChatChannelIRCChannelSetPrompt();
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-irc-channel-prompt"));
        }

    }

    private class ChatChannelIRCChannelSetPrompt extends MessagePrompt {

        @Override
        protected Prompt getNextPrompt(ConversationContext context) {
            return new ChatChannelIRCWhitelistPrompt();
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-irc-channel-valid"));
        }

    }

    private class ChatChannelIRCWhitelistPrompt extends BooleanPrompt {

        @Override
        protected Prompt acceptValidatedInput(ConversationContext context, boolean input) {
            context.setSessionData("irc_whitelist", input);
            return new ChatChannelIRCWhitelistSetPrompt();
        }

        @Override
        protected String getFailedValidationText(ConversationContext context, String invalidInput) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-irc-whitelist-invalid-boolean"));
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-irc-whitelist-prompt"));
        }

    }

    private class ChatChannelIRCWhitelistSetPrompt extends MessagePrompt {

        @Override
        protected Prompt getNextPrompt(ConversationContext context) {
            return new ChatChannelJoinedByDefaultPrompt();
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-irc-whitelist-valid"));
        }

    }

    private class ChatChannelJoinedByDefaultPrompt extends BooleanPrompt {

        @Override
        protected Prompt acceptValidatedInput(ConversationContext context, boolean input) {
            context.setSessionData("joined_by_default", input);
            return new ChatChannelJoinedByDefaultSetPrompt();
        }

        @Override
        protected String getFailedValidationText(ConversationContext context, String invalidInput) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-joined-by-default-invalid-boolean"));
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-joined-by-default-prompt"));
        }

    }

    private class ChatChannelJoinedByDefaultSetPrompt extends MessagePrompt {

        @Override
        protected Prompt getNextPrompt(ConversationContext context) {
            return new ChatChannelCreatedPrompt();
        }

        @Override
        public String getPromptText(ConversationContext context) {
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-set-joined-by-default-valid"));
        }
    }

    private class ChatChannelCreatedPrompt extends MessagePrompt {

        @Override
        protected Prompt getNextPrompt(ConversationContext context) {
            return END_OF_CONVERSATION;
        }

        @Override
        public String getPromptText(ConversationContext context) {
            BukkitChatChannelProvider chatChannelProvider = plugin.getCore().getServiceManager().getServiceProvider(BukkitChatChannelProvider.class);
            String name = (String) context.getSessionData("name");
            ChatColor color = (ChatColor) context.getSessionData("color");
            String formatString = (String) context.getSessionData("format_string");
            int radius = (int) context.getSessionData("radius");
            int clearRadius = (int) context.getSessionData("clear_radius");
            String matchPattern = (String) context.getSessionData("match_pattern");
            boolean ircEnabled = (boolean) context.getSessionData("irc_enabled");
            String ircChannel = (String) context.getSessionData("irc_channel");
            boolean ircWhitelist = (boolean) context.getSessionData("irc_whitelist");
            boolean joinedByDefault = (boolean) context.getSessionData("joined_by_default");
            BukkitChatChannel chatChannel = new BukkitChatChannel.Builder(plugin)
                    .name(name)
                    .color(color)
                    .formatString(formatString)
                    .radius(radius)
                    .clearRadius(clearRadius)
                    .matchPattern(matchPattern)
                    .ircEnabled(ircEnabled)
                    .ircChannel(ircChannel)
                    .ircWhitelist(ircWhitelist)
                    .joinedByDefault(joinedByDefault)
                    .build();
            chatChannelProvider.addChatChannel(chatChannel);
            return ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.chatchannel-create-valid"));
        }

    }

}

File: bukkit/elysium-chat-bukkit/src/main/kotlin/com/seventh_root/elysium/chat/bukkit/command/chatchannel/ChatChannelCreateCommand.kt
package com.seventh_root.elysium.chat.bukkit.command.chatchannel

import com.seventh_root.elysium.chat.bukkit.ElysiumChatBukkit
import com.seventh_root.elysium.chat.bukkit.chatchannel.BukkitChatChannel
import com.seventh_root.elysium.chat.bukkit.chatchannel.BukkitChatChannelProvider
import com.seventh_root.elysium.core.bukkit.util.ChatColorUtils
import org.bukkit.ChatColor
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.conversations.*
import org.bukkit.entity.Player
import java.awt.Color

class ChatChannelCreateCommand(private val plugin: ElysiumChatBukkit) : CommandExecutor {
    private val conversationFactory: ConversationFactory

    init {
        conversationFactory = ConversationFactory(plugin)
                .withModality(true)
                .withFirstPrompt(ChatChannelNamePrompt())
                .withEscapeSequence("cancel")
                .thatExcludesNonPlayersWithMessage(ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.not-from-console")))
                .addConversationAbandonedListener { event ->
            if (!event.gracefulExit()) {
                val conversable = event.context.forWhom
                if (conversable is Player) {
                    conversable.sendMessage(ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.operation-cancelled")))
                }
            }
        }
    }

    override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<String>): Boolean {
        if (sender is Conversable) {
            if (sender.hasPermission("elysium.chat.command.chatchannel.create")) {
                conversationFactory.buildConversation(sender).begin()
            } else {
                sender.sendMessage(ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.no-permission-chatchannel-create")))
            }
        } else {
            sender.sendMessage(ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.not-from-console")))
        }
        return true
    }

    private inner class ChatChannelNamePrompt : ValidatingPrompt() {

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-create-prompt"))
        }

        override fun isInputValid(context: ConversationContext, input: String): Boolean {
            val chatChannelProvider = plugin.core!!.serviceManager.getServiceProvider(BukkitChatChannelProvider::class.java)
            return chatChannelProvider.getChatChannel(input) == null
        }

        override fun getFailedValidationText(context: ConversationContext?, invalidInput: String?): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-name-invalid-name"))
        }

        override fun acceptValidatedInput(context: ConversationContext, input: String): Prompt {
            context.setSessionData("name", input)
            return ChatChannelNameSetPrompt()
        }

    }

    private inner class ChatChannelNameSetPrompt : MessagePrompt() {

        override fun getNextPrompt(context: ConversationContext): Prompt {
            return ChatChannelColorPrompt()
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-name-valid"))
        }

    }

    private inner class ChatChannelColorPrompt : ValidatingPrompt() {

        override fun isInputValid(context: ConversationContext, input: String): Boolean {
            try {
                ChatColor.valueOf(input.toUpperCase().replace(' ', '_'))
                return true
            } catch (exception: IllegalArgumentException) {
                return false
            }

        }

        override fun acceptValidatedInput(context: ConversationContext, input: String): Prompt {
            context.setSessionData("color", ChatColorUtils.colorFromChatColor(ChatColor.valueOf(input.toUpperCase().replace(' ', '_'))))
            return ChatChannelColorSetPrompt()
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-color-prompt"))
        }

        override fun getFailedValidationText(context: ConversationContext?, invalidInput: String?): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-color-invalid-color"))
        }

    }

    private inner class ChatChannelColorSetPrompt : MessagePrompt() {

        override fun getNextPrompt(context: ConversationContext): Prompt {
            return ChatChannelFormatStringPrompt()
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-color-valid"))
        }

    }

    private inner class ChatChannelFormatStringPrompt : StringPrompt() {

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-formatstring-prompt"))
        }

        override fun acceptInput(context: ConversationContext, input: String): Prompt {
            context.setSessionData("format_string", input)
            return ChatChannelFormatStringSetPrompt()
        }

    }

    private inner class ChatChannelFormatStringSetPrompt : MessagePrompt() {

        override fun getNextPrompt(context: ConversationContext): Prompt {
            return ChatChannelRadiusPrompt()
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-formatstring-valid"))
        }

    }

    private inner class ChatChannelRadiusPrompt : NumericPrompt() {

        override fun acceptValidatedInput(context: ConversationContext, input: Number): Prompt {
            context.setSessionData("radius", input.toInt())
            return ChatChannelRadiusSetPrompt()
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-radius-prompt"))
        }

    }

    private inner class ChatChannelRadiusSetPrompt : MessagePrompt() {

        override fun getNextPrompt(context: ConversationContext): Prompt {
            if (context.getSessionData("radius") as Int > 0) {
                return ChatChannelClearRadiusPrompt()
            } else {
                context.setSessionData("clear_radius", 0)
                return ChatChannelMatchPatternPrompt()
            }
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-radius-valid"))
        }

    }

    private inner class ChatChannelClearRadiusPrompt : NumericPrompt() {

        override fun acceptValidatedInput(context: ConversationContext, input: Number): Prompt {
            context.setSessionData("clear_radius", input.toInt())
            return ChatChannelClearRadiusSetPrompt()
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-clear-radius-prompt"))
        }

    }

    private inner class ChatChannelClearRadiusSetPrompt : MessagePrompt() {

        override fun getNextPrompt(context: ConversationContext): Prompt {
            return ChatChannelMatchPatternPrompt()
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-clear-radius-valid"))
        }

    }

    private inner class ChatChannelMatchPatternPrompt : StringPrompt() {

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-match-pattern-prompt"))
        }

        override fun acceptInput(context: ConversationContext, input: String): Prompt {
            if (!input.equals("none", ignoreCase = true))
                context.setSessionData("match_pattern", input)
            else
                context.setSessionData("match_pattern", "")
            return ChatChannelMatchPatternSetPrompt()
        }

    }

    private inner class ChatChannelMatchPatternSetPrompt : MessagePrompt() {

        override fun getNextPrompt(context: ConversationContext): Prompt {
            if (context.getSessionData("radius") as Int <= 0) {
                return ChatChannelIRCEnabledPrompt()
            } else {
                context.setSessionData("irc_enabled", false)
                context.setSessionData("irc_channel", "")
                context.setSessionData("irc_whitelist", false)
                return ChatChannelJoinedByDefaultPrompt()
            }
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-match-pattern-valid"))
        }

    }

    private inner class ChatChannelIRCEnabledPrompt : BooleanPrompt() {

        override fun getFailedValidationText(context: ConversationContext?, invalidInput: String?): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel.set-irc-enabled-invalid-boolean"))
        }

        override fun acceptValidatedInput(context: ConversationContext, input: Boolean): Prompt {
            context.setSessionData("irc_enabled", input)
            return ChatChannelIRCEnabledSetPrompt()
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-irc-enabled-prompt"))
        }

    }

    private inner class ChatChannelIRCEnabledSetPrompt : MessagePrompt() {

        override fun getNextPrompt(context: ConversationContext): Prompt {
            if (context.getSessionData("irc_enabled") as Boolean) {
                return ChatChannelIRCChannelPrompt()
            } else {
                context.setSessionData("irc_channel", "")
                context.setSessionData("irc_whitelist", false)
                return ChatChannelJoinedByDefaultPrompt()
            }
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-irc-enabled-valid"))
        }

    }

    private inner class ChatChannelIRCChannelPrompt : ValidatingPrompt() {

        override fun isInputValid(context: ConversationContext, input: String): Boolean {
            return input.matches("/([#&][^\\x07\\x2C\\s]{0,200})/".toRegex())
        }

        override fun getFailedValidationText(context: ConversationContext?, invalidInput: String?): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-irc-channel-invalid-irc-channel"))
        }

        override fun acceptValidatedInput(context: ConversationContext, input: String): Prompt {
            context.setSessionData("irc_channel", input)
            return ChatChannelIRCChannelSetPrompt()
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-irc-channel-prompt"))
        }

    }

    private inner class ChatChannelIRCChannelSetPrompt : MessagePrompt() {

        override fun getNextPrompt(context: ConversationContext): Prompt {
            return ChatChannelIRCWhitelistPrompt()
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-irc-channel-valid"))
        }

    }

    private inner class ChatChannelIRCWhitelistPrompt : BooleanPrompt() {

        override fun acceptValidatedInput(context: ConversationContext, input: Boolean): Prompt {
            context.setSessionData("irc_whitelist", input)
            return ChatChannelIRCWhitelistSetPrompt()
        }

        override fun getFailedValidationText(context: ConversationContext?, invalidInput: String?): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-irc-whitelist-invalid-boolean"))
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-irc-whitelist-prompt"))
        }

    }

    private inner class ChatChannelIRCWhitelistSetPrompt : MessagePrompt() {

        override fun getNextPrompt(context: ConversationContext): Prompt {
            return ChatChannelJoinedByDefaultPrompt()
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-irc-whitelist-valid"))
        }

    }

    private inner class ChatChannelJoinedByDefaultPrompt : BooleanPrompt() {

        override fun acceptValidatedInput(context: ConversationContext, input: Boolean): Prompt {
            context.setSessionData("joined_by_default", input)
            return ChatChannelJoinedByDefaultSetPrompt()
        }

        override fun getFailedValidationText(context: ConversationContext?, invalidInput: String?): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-joined-by-default-invalid-boolean"))
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-joined-by-default-prompt"))
        }

    }

    private inner class ChatChannelJoinedByDefaultSetPrompt : MessagePrompt() {

        override fun getNextPrompt(context: ConversationContext): Prompt {
            return ChatChannelCreatedPrompt()
        }

        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-set-joined-by-default-valid"))
        }
    }

    private inner class ChatChannelCreatedPrompt : MessagePrompt() {

        override fun getNextPrompt(context: ConversationContext): Prompt? {
            return Prompt.END_OF_CONVERSATION
        }

        override fun getPromptText(context: ConversationContext): String {
            val chatChannelProvider = plugin.core!!.serviceManager.getServiceProvider(BukkitChatChannelProvider::class.java)
            val name = context.getSessionData("name") as String
            val color = context.getSessionData("color") as Color
            val formatString = context.getSessionData("format_string") as String
            val radius = context.getSessionData("radius") as Int
            val clearRadius = context.getSessionData("clear_radius") as Int
            val matchPattern = context.getSessionData("match_pattern") as String
            val ircEnabled = context.getSessionData("irc_enabled") as Boolean
            val ircChannel = context.getSessionData("irc_channel") as String
            val ircWhitelist = context.getSessionData("irc_whitelist") as Boolean
            val joinedByDefault = context.getSessionData("joined_by_default") as Boolean
            val chatChannel = BukkitChatChannel.Builder(plugin)
                    .name(name)
                    .color(color)
                    .formatString(formatString)
                    .radius(radius)
                    .clearRadius(clearRadius)
                    .matchPattern(matchPattern)
                    .ircEnabled(ircEnabled)
                    .ircChannel(ircChannel)
                    .ircWhitelist(ircWhitelist)
                    .joinedByDefault(joinedByDefault)
                    .build()
            chatChannelProvider.addChatChannel(chatChannel)
            return ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.chatchannel-create-valid"))
        }

    }

}
